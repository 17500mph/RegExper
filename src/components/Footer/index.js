import React from 'react';
import PropTypes from 'prop-types';
import { StaticQuery, graphql } from 'gatsby';
import { withNamespaces, Trans } from 'react-i18next';

import style from './style.module.css';

const query = graphql`
  query FooterQuery {
    site {
      siteMetadata {
        buildId
      }
    }
  }
`;

export const Footer = ({ t, buildId }) => (
  <footer className={ style.footer }>
    <ul className={ style.list }>
      <li>
        <Trans>Created by <a
          href="mailto:jeff.avallone@gmail.com">Jeff Avallone</a></Trans>
      </li>
      <li>
        <Trans>Generated images licensed: <a
          href="http://creativecommons.org/licenses/by/3.0/"
          rel="license external noopener noreferrer"
          target="_blank">
          <img src="https://licensebuttons.net/l/by/3.0/80x15.png"
            alt={ t('Creative Commons CC-BY-3.0 License') } />
        </a></Trans>
      </li>
    </ul>
    <div className={ style.buildId }>
      { buildId }
    </div>
  </footer>
);

Footer.propTypes = {
  t: PropTypes.func.isRequired,
  buildId: PropTypes.string.isRequired
};

export default withNamespaces()(props => (
  <StaticQuery query={ query } render={ ({ site: { siteMetadata } }) => (
    <Footer { ...props } { ...siteMetadata } />
  ) } />
));
