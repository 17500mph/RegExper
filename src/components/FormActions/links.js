const createSvgLink = async ({ svg }) => {
  try {
    const type = 'image/svg+xml';
    const blob = new Blob([svg], { type });

    return {
      url: URL.createObjectURL(blob),
      label: 'Download SVG',
      filename: 'image.svg',
      type
    };
  }
  catch (e) {
    console.error(e); // eslint-disable-line no-console
  }
};

const createPngLink = async ({ svg, width, height }) => {
  try {
    const type = 'image/png';
    const canvas = document.createElement('canvas');
    const context = canvas.getContext('2d');
    const loader = new Image();

    loader.width = canvas.width = width * 2;
    loader.height = canvas.height = height * 2;

    await new Promise(resolve => {
      loader.onload = resolve;
      loader.src = 'data:image/svg+xml,' + encodeURIComponent(svg);
    });

    context.drawImage(loader, 0, 0, loader.width, loader.height);

    const blob = await new Promise(resolve => canvas.toBlob(resolve, type));

    return {
      url: URL.createObjectURL(blob),
      label: 'Download PNG',
      filename: 'image.png',
      type
    };
  }
  catch (e) {
    console.error(e); // eslint-disable-line no-console
  }
};

export { createSvgLink, createPngLink };
