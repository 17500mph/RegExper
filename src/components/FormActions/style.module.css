@import url('../../globals.module.css');

.actions {
  composes: inline-list with-separator-left;
  float: right;

  @media (max-width: 700px) {
    & li {
      display: block;
    }
  }

  & svg {
    width: 1em;
    height: 1em;
    margin-right: 0.5rem;
    vertical-align: text-bottom;
  }
}
