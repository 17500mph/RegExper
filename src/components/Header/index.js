import React from 'react';
import { translate, Trans } from 'react-i18next';

import style from './style.css';
import GitlabIcon from 'feather-icons/dist/icons/gitlab.svg';

import LocaleSwitcher from 'components/LocaleSwitcher';

const Header = () => (
  <header className={ style.header } data-banner={ process.env.BANNER }>
    <h1>
      <a href="/">Regexper</a>
    </h1>

    <ul>
      <li><a href="https://gitlab.com/javallone/regexper-static" rel="external noopener noreferrer" target="_blank">
        <GitlabIcon/><Trans>Source on GitLab</Trans>
      </a></li>
      <li>
        <a href="/privacy.html"><Trans>Privacy Policy</Trans></a>
      </li>
      <li><LocaleSwitcher /></li>
    </ul>
  </header>
);

export default translate()(Header);
export { Header };
