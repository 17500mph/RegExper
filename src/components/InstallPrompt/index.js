import React from 'react';
import { withTranslation, Trans } from 'react-i18next';

class InstallPrompt extends React.PureComponent {
  state = {
    installPrompt: null
  }

  componentDidMount() {
    window.addEventListener('beforeinstallprompt', this.handleInstallPrompt);
  }

  componentWillUnmount() {
    window.removeEventListener('beforeinstallprompt', this.handleInstallPrompt);
  }

  handleInstallPrompt = event => {
    this.setState({
      installPrompt: event
    });
  }

  handleInstall = async event => {
    event.preventDefault();

    const { installPrompt } = this.state;

    try {
      installPrompt.prompt();
      await installPrompt.userChoice;
    }
    catch {
      // User cancelled install
    }

    this.setState({ installPrompt: null });
  }

  render() {
    const { installPrompt } = this.state;

    if (!installPrompt) {
      return null;
    }

    return <a href="#install" onClick={ this.handleInstall }>
      <Trans>Add to Home Screen</Trans>
    </a>;
  }
}

export { InstallPrompt };
export default withTranslation()(InstallPrompt);
