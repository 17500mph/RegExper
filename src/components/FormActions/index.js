import React from 'react';
import PropTypes from 'prop-types';
import { withNamespaces, Trans } from 'react-i18next';

import DownloadIcon from 'react-feather/dist/icons/download';
import LinkIcon from 'react-feather/dist/icons/link';

import style from './style.module.css';

import { createPngLink, createSvgLink } from './links';

class FormActions extends React.PureComponent {
  static propTypes = {
    permalinkUrl: PropTypes.string,
    imageDetails: PropTypes.shape({
      svg: PropTypes.string,
      width: PropTypes.number,
      height: PropTypes.number
    }),
    t: PropTypes.func.isRequired
  }

  state = {
    svgLink: null,
    pngLink: null
  }

  componentDidMount() {
    const { imageDetails } = this.props;

    if (imageDetails && imageDetails.svg) {
      this.generateDownloadLinks();
    }
  }

  componentDidUpdate(prevProps) {
    const { imageDetails } = this.props;
    const { imageDetails: prevImageDetails } = prevProps;

    if (!imageDetails) {
      this.setState({ svgLink: null, pngLink: null });
      return;
    }

    if (!prevImageDetails) {
      this.generateDownloadLinks();
      return;
    }

    if (imageDetails.svg !== prevImageDetails.svg
      || imageDetails.width !== prevImageDetails.width
      || imageDetails.height !== prevImageDetails.height) {
      this.generateDownloadLinks();
      return;
    }
  }

  async generateDownloadLinks() {
    const { imageDetails: { svg, width, height } } = this.props;

    this.setState({
      svgLink: await createSvgLink({ svg }),
      pngLink: await createPngLink({ svg, width, height })
    });
  }

  downloadLink({ url, filename, type, label }) {
    const { t } = this.props;

    return <li>
      <a href={ url } download={ filename } type={ type }>
        <DownloadIcon />{ t(label) }
      </a>
    </li>;
  }

  render() {
    const {
      permalinkUrl
    } = this.props;
    const {
      svgLink,
      pngLink
    } = this.state;

    return <ul className={ style.actions }>
      { pngLink && this.downloadLink(pngLink) }
      { svgLink && this.downloadLink(svgLink) }
      { permalinkUrl && <li>
        <a href={ permalinkUrl }><LinkIcon /><Trans>Permalink</Trans></a>
      </li> }
    </ul>;
  }
}

export { FormActions };
export default withNamespaces()(FormActions);
