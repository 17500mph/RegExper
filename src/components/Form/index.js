import React from 'react';

import ExpandIcon from 'react-feather/dist/icons/chevrons-down';

import style from './style.module.css';

import AppContext from 'components/App/context';
import FormActions from 'components/FormActions';

const syntaxList = [
  { id: 'js', label: 'JavaScript' },
  { id: 'pcre', label: 'PCRE' }
];

class Form extends React.PureComponent {
  static contextType = AppContext

  state = {
    expr: this.context.expr,
    syntax: this.context.syntax
  }

  componentDidUpdate() {
    if (this.state.expr === undefined && this.state.syntax === undefined) {
      this.setState({
        syntax: this.context.syntax,
        expr: this.context.expr
      });
    }
  }

  handleSubmit = event => {
    event.preventDefault();

    const { expr, syntax } = this.state;

    this.context.renderExpr({ expr, syntax });
  }

  handleKeyPress = event => {
    if (event.charCode === 13 && event.shiftKey) {
      this.handleSubmit(event);
    }
  }

  handleChange = event => this.setState({
    [event.target.name]: event.target.value
  })

  render() {
    const { expr, syntax } = this.state;

    return <div className={ style.form }>
      <form onSubmit={ this.handleSubmit }>
        <textarea
          name="expr"
          value={ expr }
          onKeyPress={ this.handleKeyPress }
          onChange={ this.handleChange }
          autoFocus
          placeholder="Enter regular expression to display"></textarea>
        <button type="submit">Display</button>
        <div className={ style.select }>
          <select
            name="syntax"
            value={ syntax }
            onChange={ this.handleChange } >
            { syntaxList.map(({ id, label }) => (
              <option value={ id } key={ id }>{ label }</option>
            )) }
          </select>
          <ExpandIcon />
        </div>
        <FormActions />
      </form>
    </div>;
  }
}

export default Form;
