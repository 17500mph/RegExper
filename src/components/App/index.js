import React from 'react';
import PropTypes from 'prop-types';
import * as Sentry from '@sentry/browser';
import URLSearchParams from '@ungap/url-search-params';

import Form from 'components/Form';
import FormActions from 'components/FormActions';
import Loader from 'components/Loader';
import Message from 'components/Message';

class App extends React.PureComponent {
  state = {
    loading: false,
    loadingError: null,
    Render: null
  }

  componentDidMount() {
    if (this.props.expr) {
      this.handleRender();
    }
  }

  componentDidUpdate(prevProps) {
    const { syntax, expr } = this.props;

    if (syntax !== prevProps.syntax || expr !== prevProps.expr) {
      this.handleRender();
    }
  }

  handleSubmit = ({ syntax, expr }) => {
    if (expr) {
      document.location.hash = new URLSearchParams({
        syntax,
        expr
      }).toString();
    }
  }

  handleRender = async () => {
    const { syntax, expr } = this.props;

    this.setState({
      loading: false,
      loadingError: null,
      Render: null
    });

    if (!expr) {
      return;
    }

    this.setState({
      loading: true
    });

    try {
      const Render = await import(
        /* webpackChunkName: "render-[index]" */
        `syntax/${ syntax }`
      );

      // HACK: Fake loading time
      await new Promise(resolve => setTimeout(resolve, 2000));

      this.setState({
        loading: false,
        Render: Render.default
      });
    }
    catch (e) {
      Sentry.withScope(scope => {
        scope.setExtra('syntax', syntax);
        Sentry.captureException(e);
      });
      this.setState({
        loading: false,
        loadingError: e
      });
      throw e;
    }
  }

  handleRetry = event => {
    event.preventDefault();
    this.handleRender();
  }

  handleSvg = imageDetails => this.setState({ imageDetails });

  render() {
    const {
      syntax,
      expr,
      permalinkUrl,
      syntaxList
    } = this.props;
    const {
      loading,
      loadingError,
      Render,
      imageDetails
    } = this.state;

    const formProps = {
      onSubmit: this.handleSubmit,
      syntax,
      expr,
      syntaxList
    };
    const actionProps = {
      imageDetails,
      permalinkUrl
    };
    const renderProps = {
      onRender: this.handleSvg,
      expr
    };

    return <>
      <Form { ...formProps }>
        { Render && <FormActions { ...actionProps } /> }
      </Form>

      { loading && <Loader /> }

      { loadingError && <Message type="error" heading="Render Failure">
        <p>An error occurred while rendering the regular expression.</p>
        <a href="#retry" onClick={ this.handleRetry }>Retry</a>
      </Message> }

      { Render && <Render { ...renderProps } /> }
    </>;
  }
}

App.propTypes = {
  syntax: PropTypes.string,
  expr: PropTypes.string,
  permalinkUrl: PropTypes.string,
  syntaxList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    label: PropTypes.string
  }))
};

export default App;
