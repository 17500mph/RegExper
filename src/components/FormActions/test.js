jest.mock('./links');
jest.mock('react-feather/dist/icons/download', () =>
  require('__mocks__/component-mock')(
    'react-feather/dist/icons/download'));
jest.mock('react-feather/dist/icons/link', () =>
  require('__mocks__/component-mock')(
    'react-feather/dist/icons/link'));

import React from 'react';
import { render } from 'react-testing-library';

import { mockT } from 'i18n';
import { FormActions } from 'components/FormActions';
import { createPngLink, createSvgLink } from './links';

createPngLink.mockResolvedValue({
  url: 'http://example.com/image.png',
  filename: 'image.png',
  type: 'image/png',
  label: 'Example PNG Link'
});
createSvgLink.mockResolvedValue({
  url: 'http://example.com/image.svg',
  filename: 'image.svg',
  type: 'image/svg+xml',
  label: 'Example SVG Link'
});

describe('FormActions', () => {
  test('rendering', () => {
    const { asFragment } = render(
      <FormActions t={ mockT } />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering with a permalink', () => {
    const { asFragment } = render(
      <FormActions permalinkUrl="http://example.com" t={ mockT } />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering download links', async () => {
    const imageDetails = {
      svg: 'test image',
      width: 10,
      height: 20
    };

    const { asFragment } = render(
      <FormActions imageDetails={ imageDetails } t={ mockT } />
    );

    // Give a beat for mocked promises to resolve
    await new Promise(resolve => setTimeout(resolve));

    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering download links with data after mounting', async () => {
    const { asFragment, rerender } = render(
      <FormActions t={ mockT } />
    );

    rerender(
      <FormActions permalinkUrl="http://example.com" t={ mockT } />
    );

    expect(asFragment()).toMatchSnapshot();

    rerender(
      <FormActions
        permalinkUrl="http://example.com"
        imageDetails={ { svg: 'test-image' } }
        t={ mockT } />
    );

    // Give a beat for mocked promises to resolve
    await new Promise(resolve => setTimeout(resolve));

    expect(asFragment()).toMatchSnapshot();

    rerender(
      <FormActions
        permalinkUrl="http://example.com"
        imageDetails={ { svg: 'test-image', width: 10, height: 20 } }
        t={ mockT } />
    );

    // Give a beat for mocked promises to resolve
    await new Promise(resolve => setTimeout(resolve));

    expect(asFragment()).toMatchSnapshot();
  });
});
