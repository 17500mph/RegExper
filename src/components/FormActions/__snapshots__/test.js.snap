// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormActions rendering 1`] = `
<DocumentFragment>
  <ul
    class="actions"
  />
</DocumentFragment>
`;

exports[`FormActions rendering download links 1`] = `
<DocumentFragment>
  <ul
    class="actions"
  >
    <li>
      <a
        download="image.png"
        href="http://example.com/image.png"
        type="image/png"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example PNG Link)
      </a>
    </li>
    <li>
      <a
        download="image.svg"
        href="http://example.com/image.svg"
        type="image/svg+xml"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example SVG Link)
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`FormActions rendering download links with data after mounting 1`] = `
<DocumentFragment>
  <ul
    class="actions"
  >
    <li>
      <a
        href="http://example.com"
      >
        <span
          data-component="Link"
          data-props="{}"
        />
        <span
          data-component="Trans"
          data-props="{}"
        >
          Permalink
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`FormActions rendering download links with data after mounting 2`] = `
<DocumentFragment>
  <ul
    class="actions"
  >
    <li>
      <a
        download="image.png"
        href="http://example.com/image.png"
        type="image/png"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example PNG Link)
      </a>
    </li>
    <li>
      <a
        download="image.svg"
        href="http://example.com/image.svg"
        type="image/svg+xml"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example SVG Link)
      </a>
    </li>
    <li>
      <a
        href="http://example.com"
      >
        <span
          data-component="Link"
          data-props="{}"
        />
        <span
          data-component="Trans"
          data-props="{}"
        >
          Permalink
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`FormActions rendering download links with data after mounting 3`] = `
<DocumentFragment>
  <ul
    class="actions"
  >
    <li>
      <a
        download="image.png"
        href="http://example.com/image.png"
        type="image/png"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example PNG Link)
      </a>
    </li>
    <li>
      <a
        download="image.svg"
        href="http://example.com/image.svg"
        type="image/svg+xml"
      >
        <span
          data-component="Download"
          data-props="{}"
        />
        TRANSLATE(Example SVG Link)
      </a>
    </li>
    <li>
      <a
        href="http://example.com"
      >
        <span
          data-component="Link"
          data-props="{}"
        />
        <span
          data-component="Trans"
          data-props="{}"
        >
          Permalink
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`FormActions rendering with a permalink 1`] = `
<DocumentFragment>
  <ul
    class="actions"
  >
    <li>
      <a
        href="http://example.com"
      >
        <span
          data-component="Link"
          data-props="{}"
        />
        <span
          data-component="Trans"
          data-props="{}"
        >
          Permalink
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
